doc/devdocs/locks.rst

```diff
@@ -134,7 +103,31 @@ Type application : typecache lock
 
 Module serializer : toplevel lock
 
-JIT : codegen lock
+JIT & type-inference : codegen lock
+
+MethodInstance updates : codegen lock
+    - these fields are generally lazy initialized, using the test-and-test-and-set pattern
+    - Set at construction, immutable:
+      + specTypes
+      + sparam_vals
+      + def
+    - set by jl_type_infer (holding codegen lock)
+      + rettype
+      + inferred
+      + these can be reset, see jl_set_lambda_rettype for that logic as it needs to keep functionObjectsDecls in sync
+    - inInference
+      + optimization to quickly avoid recurring into jl_type_infer while it is already running
+      + actual state (of inferred then fptr) is protected by codegen lock
+    - jlcall_api and fptr, unspecialized_ducttape
+      + these transition once, from NULL to a value, while the codegen lock is held
+    - functionObjectsDecls
+      + these can transition multiple times, but only while the codegen lock is held
+      + it is valid to use old version of this, or wait for new versions of this,
+        so races are benign, as long as the code is careful not to reference other data in the lambda (such as rettype)
```

Comment:
this is invalid rst `/Users/travis/build/JuliaLang/julia/doc/devdocs/locks.rst:126: ERROR: Unexpected indentation.`